ffi/src/lib.rs

```diff
@@ -313,162 +568,365 @@ fn unwrap_conversion<T>(value: Option<T>, expected_type: ValueType) -> T {
 #[no_mangle]
 pub unsafe extern "C" fn typed_value_as_timestamp(typed_value: *mut TypedValue) ->  c_longlong {
     let typed_value = Box::from_raw(typed_value);
     unwrap_conversion(typed_value.into_timestamp(), ValueType::Instant)
 }
 
-//as_string
+/// Consumes a `TypedValue` and returns the value as a `String`.
+///
+/// # Panics
+///
+/// If the `ValueType` of the `TypedValue` is not `ValueType::String`.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `TypedValue` is not dangling .
 #[no_mangle]
 pub unsafe extern "C" fn typed_value_as_string(typed_value: *mut TypedValue) ->  *const c_char {
     let typed_value = Box::from_raw(typed_value);
     c_char_from_rc(unwrap_conversion(typed_value.into_string(), ValueType::String))
 }
 
-//as_uuid
+/// Consumes a `TypedValue` and returns the value as a UUID byte slice of length 16.
+///
+/// # Panics
+///
+/// If the `ValueType` of the `TypedValue` is not `ValueType::Uuid`.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `TypedValue` is not dangling .
 #[no_mangle]
 pub unsafe extern "C" fn typed_value_as_uuid(typed_value: *mut TypedValue) ->  *mut [u8; 16] {
     let typed_value = Box::from_raw(typed_value);
     let value = unwrap_conversion(typed_value.into_uuid(), ValueType::Uuid);
     Box::into_raw(Box::new(*value.as_bytes()))
 }
 
-//value_type
+/// Returns the `ValueType` of this `TypedValue`.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `TypedValue` is not dangling .
 #[no_mangle]
 pub unsafe extern "C" fn typed_value_value_type(typed_value: *mut TypedValue) ->  ValueType {
     let typed_value = &*typed_value;
     typed_value.value_type()
 }
 
+/// Returns the value at the provided `index` as a `Vec<ValueType>`.
+/// If there is no value present at the `index`, a null pointer is returned.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `Vec<Vec<TypedValue>>` is not dangling .
+///
+/// Callers are responsible for managing the memory for the return value.
+/// A destructor `typed_value_result_set_destroy` is provided for releasing the memory for this
+/// pointer type.
 #[no_mangle]
 pub unsafe extern "C" fn row_at_index(rows: *mut Vec<Vec<TypedValue>>, index: c_int) ->  *mut Vec<TypedValue> {
     let result = &*rows;
     result.get(index as usize).map_or(std::ptr::null_mut(), |v| Box::into_raw(Box::new(v.clone())))
 }
 
+/// Consumes the `Vec<Vec<TypedValue>>` and returns an iterator over the values.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `Vec<Vec<TypedValue>>` is not dangling .
+///
+/// Callers are responsible for managing the memory for the return value.
+/// A destructor `typed_value_result_set_iter_destroy` is provided for releasing the memory for this
+/// pointer type.
 #[no_mangle]
 pub unsafe extern "C" fn rows_iter(rows: *mut Vec<Vec<TypedValue>>) ->  *mut TypedValueListIterator {
```

Comment:
Again, `into_rows_iter`?

Reply:
Convention with FFI states that you start the function name with the type of the pointer that the function acts on. With that in mind, this should probably be `typed_value_result_into_iter` or something. Will rename.

Reply:
Your choice.